src/net.h

```diff
@@ -64,6 +64,8 @@ static const unsigned int MAX_PROTOCOL_MESSAGE_LENGTH = 4 * 1000 * 1000;
 static const unsigned int MAX_SUBVERSION_LENGTH = 256;
 /** Maximum number of automatic outgoing nodes over which we'll relay everything (blocks, tx, addrs, etc) */
 static const int MAX_OUTBOUND_FULL_RELAY_CONNECTIONS = 8;
+/** Maximum number of automatic fullrbf peers */
+static const int MAX_FULLRBF_RELAY_CONNECTIONS = 1;
```

Comment:
I believe this should be higher. With just ~1 full-RBF peer per node, a full-RBF tx is propagating in a network that is "chain-like". At each propagation step, the # of new nodes receiving the tx will be approximately one.

You should set this higher, to 4 or so, to ensure that propagation is "exponential" in nature, with each propagating step reaching more and more nodes.

Reply:
You also have to take into account that each node makes 8 automatic outgoing connections that might by chance connect to more full-RBF peers. This chance becomes higher if more peers run full-RBF. So it really comes down to how small of a minority of supporting nodes we require for full-RBF to still work reliably. 

Also note that increasing automatic outbound slots too much could be problematic if full-rbf was adopted by many nodes, because there might not be sufficient inbound capacity. Studies like https://arxiv.org/pdf/2108.00815v2.pdf (Fig. 3) suggest that a significant number of nodes already operate close to their maximum inbound capacity today.

Reply:
> You also have to take into account that each node makes 8 automatic outgoing connections that might by chance connect to more full-RBF peers.

The chance is pretty low at small %'s of full-RBF nodes. I know none of my nodes were connected to RBF-advertising peers by accident without this patch.

> Also note that increasing automatic outbound slots too much could be problematic if full-rbf was adopted by many nodes

I think that's quite unlikely. I highly doubt more than a few % of nodes would enable this manually. Meanwhile, once a few % of nodes and miners have, any remaining arguments against enabling full-rbf by default are certainly bunk, as getting non-opt-in replacements mined will be easy.

Reply:
I think the chances to be connected to a full-rbf subgraph by relying only on the 8 automatic outgoing connections sounds really low, even assuming you're a listening node. If you model something with ~60k nodes, ~125 connection slots and 100 full-rbf nodes across the network, the success probability is like what ~1/60 ? This could explain the observed lack of RBF-advertising peers connections.

Thinking back to the model laid out above, rather to define success as the existence of a full-rbf route between two full-rbf supporting nodes another success definition could be the propagation of a full-rbf tx among the quasi-unanimity of the full-rbf peers. Therefore to achieve success we should aim to eradicate disconnected graphs of full-rbf peers. The exact value of `MAX_FULLRBF_RELAY_CONNECTIONS` would be function of the set size of peers we assume enabling full-rbf by option.

I think the concern about increased consumption of inbound capacity is qualified, however I would say best we can do is giving us a slot budget. I.e assuming a set size of full-rbf peers, a number of new automatic outbound connections and an occupation rate of the inbound slots of the listening peers and gauge if the result is acceptable.

Reply:
> The exact value of `MAX_FULLRBF_RELAY_CONNECTIONS` would be function of the set size of peers we assume enabling full-rbf by option.

No, that's not how this works. Since we're doing preferential peering, the total number of peers with full-rbf on the network isn't relevant. We simply need a figure high enough that a sufficient number of full-rbf-signalling peers we connect to are in fact, actuall full-rbf peers that likewise have preferential peering. Second, by "sufficient number", we need something greater than 1 so we end up with a well-connected group rather than disconnected chains. 4+ should be fine.

So the actual figure we care about is what % of peers with the full-rbf service bit set are running this code (or something like it)? There are a handful of nodes running Knots that advertise the service bit but don't do preferential peering. Once a reasonable number of people (a few dozen) use this flag, they probably won't matter as AFAICT there's less than half a dozen such Knots-running nodes.

Reply:
> We simply need a figure high enough that a sufficient number of full-rbf-signalling peers we connect to are in fact, actuall full-rbf peers that likewise have preferential peering. Second, by "sufficient number", we need something greater than 1 so we end up with a well-connected group rather than disconnected chains. 4+ should be fine.

Yes, I think it would be interesting to have a consistent propagation model to determine what exact "sufficient number" would fit to ensure we have well-connected full-rbf group rather than disconnected chains. I wonder if there have been such models established in the past Core discussions about p2p modifications or extensions. Otherwise landing on the number 4+ sounds really found by empiricism and historical observations ? (not a bad method we should just be aware of the limits)